pxr/base/tf/delegatedCountPtr.h

```diff
@@ -0,0 +1,271 @@
+    TfDelegatedCountPtr() noexcept = default;
+
+    /// Creates a new pointer without retaining
+    /// \sa TfDelegatedCountDoNotIncrementTag
+    TfDelegatedCountPtr(TfDelegatedCountDoNotIncrementTagType,
+                        RawPtrType rawPointer) noexcept :
+        _pointer{rawPointer} {
+    }
+
+    /// Creates a new retained pointer
+    ///
+    /// Retains only if the pointer is valid.
+    /// \sa TfDelegatedCountIncrementTag
+    TfDelegatedCountPtr(TfDelegatedCountIncrementTagType,
+                        RawPtrType rawPointer)
+    noexcept(IncrementIsNoExcept()) :
+        _pointer{rawPointer} {
+        _IncrementIfValid();
+    }
+
+    /// Constructs by copying the pointer from `ptr`.
+    ///
+    /// Retains only if the pointer is valid.
+    TfDelegatedCountPtr(const TfDelegatedCountPtr& ptr)
+    noexcept(IncrementIsNoExcept()) :
+        _pointer{ptr.get()} {
+        _IncrementIfValid();
+    }
+
+    /// Constructs by copying a convertible pointer from `ptr`.
+    ///
+    /// This is necessary in part to allow a `const` `ValueType` to be
+    /// constructed from from a non-`const` `ValueType`.
+    ///
+    /// Retains only if the pointer is valid.
+    template <typename OtherType>
+    explicit TfDelegatedCountPtr(
+        const TfDelegatedCountPtr<OtherType>& ptr,
+        std::enable_if_t<_IsPtrConvertible<OtherType>::value, int> = 0)
+    noexcept(IncrementIsNoExcept()) :
+        _pointer(ptr.get()) {
+        _IncrementIfValid();
+    }
+
+    /// Constructs by moving the pointer from `ptr`.
+    ///
+    /// `ptr` will be reset to its default state (ie. `nullptr`).
+    TfDelegatedCountPtr(TfDelegatedCountPtr&& ptr) noexcept :
+        _pointer(ptr.get()) {
+        ptr._pointer = nullptr;
+    }
+
+    /// Assigns by copying the pointer from `ptr`.
+    ///
+    /// Retains only if the pointer is valid.
+    TfDelegatedCountPtr& operator=(const TfDelegatedCountPtr& ptr)
+    noexcept(IncrementAndDecrementAreNoExcept()) {
+        // Implement copy assigment in terms of move assignment
+        return (*this = TfDelegatedCountPtr{ptr});
+    }
+
+    /// Assigns by copying a convertible pointer from `ptr`.
+    ///
+    /// This is necessary in part to allow a `const` `ValueType` to be
+    /// assigned from from a non-`const` `ValueType`.
+    ///
+    /// Retains only if the pointer is valid.
+    template <typename OtherType>
+    TfDelegatedCountPtr& operator=(
+        const TfDelegatedCountPtr<OtherType>& ptr)
+    noexcept(IncrementAndDecrementAreNoExcept()) {
+        static_assert(_IsPtrConvertible<OtherType>::value);
+        // Implement copy assigment in terms of move assignment
+        return (*this = TfDelegatedCountPtr{ptr});
+    }
+
+    /// Assigns by moving the pointer from `ptr`.
+    ///
+    /// `ptr` will be reset to its default state (ie. `nullptr`).
+    TfDelegatedCountPtr& operator=(TfDelegatedCountPtr&& ptr)
+    noexcept(DecrementIsNoExcept()) {
```

Comment:
ok if I had a `foo` that was a `TfDelegatedCountPtr<Foo>` and I did:
```
foo = std::move(foo);
```
That would blow up, right?

Reply:
There's a test case for this specifically.  It doesn't blow up.  `foo` is left in a valid state. See `TestMovingSelf()`.